### DISCUSSION TITLE
Kalman reset through cflib

### DISCUSSION THREAD

**Discussion by NageshEranki** [bitcraze/discussions#1431] [Q&A]
Hi 

I am using the example script to integrate the qualisys system with the Crazyflie 2.1 [link](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/mocap/qualisys_hl_commander.py)

I see the following lines of code to reset the state estimated by the EKF:

```
def reset_estimator(cf):
    cf.param.set_value('kalman.resetEstimation', '1')
    time.sleep(0.1)
    cf.param.set_value('kalman.resetEstimation', '0')
```

I don't understand why we need to reset the parameter back to '0' because when we first set it to '1', the firmware will reset the state and change the parameter back to '0' anyway. I found an earlier discussion on the old forum on this: [thread](https://forum.bitcraze.io/viewtopic.php?t=2746) but what I don't understand is that if the param is set back to '1', then every call to ```cf.param.set_value('kalman.resetEstimation', '1')``` triggers a 'rising edge' anyway. Am I missing something?

Thanks,
Nagesh

**Comment by ataffanel** [comment]
Hi, I am most likely the one that have written that so I can try to explain.

You are right, when setting the value to 1 the current implementation on the lib always sends a packet with a 1 that triggers the reset in that case since the firmware resets the value automatically.

Though, this is not how the parameter subsystem was originally though and designed: the parameter subsystem was designed for values that can be changed from the ground and are never changed by the Crazyflie firmware. This means that values can be cached and that it would be completely within spec for the client to cache all values and to not send another "1" is the current value is already "1". Hence this code that makes sure that the parameter flips from 0 to 1 and then back to 0.

We have since implemented an api to allow the firmware to set a parameter which sends a packet to the ground updating the ground's cache, if that is used in the firmware, it would cause no problem to only set the value to 1 from the ground.
